app/models/concerns/previewing.rb

```diff
@@ -0,0 +1,83 @@
+module Previewing
+  extend ActiveSupport::Concern
+
+  included do
+    def self.collection_preview(saved, params)
+      child_args(saved, params).map do |obj, attrs|
+        if obj
+          obj.preview(attrs)
+        else
+          self.new(attrs)
+        end
+      end
+    end
+
+    private
+
+    def self.child_args(saved, child_attrs)
+      if saved.size > child_attrs.size
+        saved.zip(child_attrs)
+      else
+        child_attrs.zip(saved).map(&:reverse!)
+      end
+    end
+  end
+
+  # Returns an unsaved object with updated attributes, along with any
+  # child objects and their updated attributes.
+  # Note that the associations are not assigned, so parent.children will be
+  # empty
+  #
+  # e.g. for Topics, the return format is
+  # [ topic_preview, [beginner_lesson, intermediate_lesson, ...]
```

Comment:
They can but there's some strange behavior -- for instance, `Topic#preview` assigns the association `admin_lessons`, not `lessons`, but when trying to set `@topic.lessons = @topic.admin_lessons` in the controller, the new attributes are lost (`instructor_students_ratio` in this example):
```
(byebug) @topic.admin_lessons
#<ActiveRecord::Associations::CollectionProxy [#<Lesson id: 63, topic_id: 99, duration: #<Duration:0x00007f77ac740998 @hours=0, @minutes=0>, body: "lesson body", created_at: "2018-10-18 01:01:08", updated_at: "2018-10-18 01:01:08", objective: "", level_id: 0, pdf: "A topic - Beginning.pdf", instructor_students_ratio: "1:4", prerequisites: "", notes: "", suggested_materials: "", published: true, recommended_reading: "", relevant_articles: "">]>
(byebug) next

[33, 42] in /home/sky/sec/app/controllers/topics_controller.rb
   33:     @preview = true
   34:     @topic = Topic.friendly.find(params[:id])
   35:     @topic = @topic.preview(preview_params.to_h)
   36:     byebug
   37:     @topic.lessons = @topic.admin_lessons.select(&:published)
=> 38:     @lesson = @topic.lessons[0] unless @topic.description?
   39:     @preview_params = { topic: preview_params.to_h }
   40:     og_object @topic
   41:     breadcrumbs @topic.name
   42:     render "show"
(byebug) @topic.lessons
#<ActiveRecord::Associations::CollectionProxy [#<Lesson id: 63, topic_id: 99, duration: #<Duration:0x00007f77ad605fc8 @hours=0, @minutes=0>, body: "lesson body", created_at: "2018-10-18 01:01:08", updated_at: "2018-10-18 01:01:08", objective: nil, level_id: 0, pdf: "A topic - Beginning.pdf", instructor_students_ratio: nil, prerequisites: nil, notes: nil, suggested_materials: nil, published: true, recommended_reading: nil, relevant_articles: nil>]>
```
I was able to fix this by setting `@topic.lessons` to an empty array before re-assigning, but this ended up breaking the new update-from-preview integration test somehow (`#<ActiveRecord::RecordNotFound: Couldn't find Lesson with ID=70 for Topic with ID=106>` on sending the `PATCH`). I'm not really sure what's happening there -- I'll separate out this change into a separate commit for now, because it does simplify `#preview` by quite a bit.

Reply:
That does seem odd. I agree that the changes to the controller are nice. Want to take a look together the next time you're in the office? I'd be curious to figure out what's going on there! Or if you'd rather move on from this, I'm okay with sticking with the original format for the return value - I don't think this change is critical.